graphql/executor/executor.go

```diff
@@ -11,115 +11,58 @@ import (
-			previous := e.responseMiddleware
-			e.responseMiddleware = func(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
-				return p.InterceptResponse(ctx, func(ctx context.Context) *graphql.Response {
-					return previous(ctx, next)
-				})
-			}
-		}
-
-		if p, ok := p.(graphql.FieldInterceptor); ok {
-			previous := e.fieldMiddleware
-			e.fieldMiddleware = func(ctx context.Context, next graphql.Resolver) (res interface{}, err error) {
-				return p.InterceptField(ctx, func(ctx context.Context) (res interface{}, err error) {
-					return previous(ctx, next)
-				})
-			}
-		}
-	}
-
-	for _, p := range s.extensions {
-		if p, ok := p.(graphql.OperationParameterMutator); ok {
-			e.operationParameterMutators = append(e.operationParameterMutators, p)
-		}
+	errorPresenter graphql.ErrorPresenterFunc
+	recoverFunc    graphql.RecoverFunc
+	queryCache     graphql.Cache
+}
 
-		if p, ok := p.(graphql.OperationContextMutator); ok {
-			e.operationContextMutators = append(e.operationContextMutators, p)
-		}
+var _ graphql.GraphExecutor = &Executor{}
 
+// New creates a new Executor with the given schema, and a default error and
+// recovery callbacks, and no query cache or extensions.
+func New(es graphql.ExecutableSchema) *Executor {
+	e := &Executor{
+		es:             es,
+		errorPresenter: graphql.DefaultErrorPresenter,
+		recoverFunc:    graphql.DefaultRecover,
+		queryCache:     graphql.NoCache{},
 	}
-
+	e.setExtensions()
 	return e
 }
 
-func (e executor) DispatchOperation(ctx context.Context, rc *graphql.OperationContext) (h graphql.ResponseHandler, resctx context.Context) {
-	ctx = graphql.WithOperationContext(ctx, rc)
-
-	var innerCtx context.Context
-	res := e.operationMiddleware(ctx, func(ctx context.Context) graphql.ResponseHandler {
-		innerCtx = ctx
-
-		tmpResponseContext := graphql.WithResponseContext(ctx, e.server.errorPresenter, e.server.recoverFunc)
-		responses := e.server.es.Exec(tmpResponseContext)
-		if errs := graphql.GetErrors(tmpResponseContext); errs != nil {
-			return graphql.OneShot(&graphql.Response{Errors: errs})
-		}
-
-		return func(ctx context.Context) *graphql.Response {
-			ctx = graphql.WithResponseContext(ctx, e.server.errorPresenter, e.server.recoverFunc)
-			resp := e.responseMiddleware(ctx, func(ctx context.Context) *graphql.Response {
-				resp := responses(ctx)
-				if resp == nil {
-					return nil
-				}
-				resp.Errors = append(resp.Errors, graphql.GetErrors(ctx)...)
-				resp.Extensions = graphql.GetExtensions(ctx)
-				return resp
-			})
-			if resp == nil {
-				return nil
-			}
+// Exec returns a graphql response from the given params.
+func (e *Executor) Exec(ctx context.Context, params *graphql.RawParams) *graphql.Response {
+	ctx = graphql.StartOperationTrace(ctx)
+	now := graphql.Now()
+	params.ReadTime = graphql.TraceTiming{
+		Start: now,
+		End:   now,
+	}
 
-			return resp
-		}
-	})
+	rc, err := e.CreateOperationContext(ctx, params)
```

Comment:
This already happens in `DispatchOperation`, In fact, If `GetStartTime()` fell back to Now() then Exec could be removed again in favour of calling `DispatchOperation` directly?

I don't like widening this interface for a fairly minor convenience method, and one that wont support subscriptions.

Reply:
I ended up just removing the convenience method since it was only used in tests.